memory/src/main/scala/filodb.memory/data/OffheapLFSortedIDMap.scala

```diff
@@ -246,22 +294,151 @@ class OffheapLFSortedIDMapReader(memFactory: MemFactory, holderKlass: Class[_ <:
+    // Spin-lock implementation.
+
+    var lockState = 0
+
+    // First set the high bit, to signal an exclusive lock request.
+
+    var done = false
+    do {
+      lockState = UnsafeUtils.getIntVolatile(inst, lockStateOffset)
+      if (lockState < 0) {
+        // Wait for exclusive lock to be released.
+        Thread.`yield`
+      } else if (UnsafeUtils.unsafe.compareAndSwapInt(inst, lockStateOffset, lockState, lockState | 0x80000000)) {
+        if (lockState == 0) {
+          return true
+        }
+        // Scala doesn't support a simple break statement for some reason.
+        done = true
+      }
+    } while (!done)
+
+    // Wait for shared lock owners to release the lock.
+
+    val endNanos = System.nanoTime + timeoutNanos
+
+    do {
+      Thread.`yield`
+      lockState = UnsafeUtils.getIntVolatile(inst, lockStateOffset)
+      if ((lockState & 0x7fffffff) == 0) {
+        return true
+      }
+    } while (System.nanoTime() < endNanos)
+
+    // Timed out. Release the exclusive lock request signal and yield (to permit shared access again).
+
+    while(!UnsafeUtils.unsafe.compareAndSwapInt(inst, lockStateOffset, lockState, lockState & 0x7fffffff)) {
+      lockState = UnsafeUtils.getIntVolatile(inst, lockStateOffset)
+    }
+
+    Thread.`yield`
+    return false
+  }
+
+  /**
+   * Release an acquired exclusive lock.
+   */
+  def releaseExclusive(inst: MapHolder): Unit = {
+    UnsafeUtils.setIntVolatile(inst, lockStateOffset, 0)
+  }
+
+  /**
+   * Acquire shared access to this map, spinning if necessary.
+   */
+  def acquireShared(inst: MapHolder): Unit = {
+    // Spin-lock implementation.
+
+    var lockState = 0
+
+    while (true) {
+      lockState = UnsafeUtils.getIntVolatile(inst, lockStateOffset)
+      if (lockState < 0) {
+        // Wait for exclusive lock to be released.
+        Thread.`yield`
+      } else if (UnsafeUtils.unsafe.compareAndSwapInt(inst, lockStateOffset, lockState, lockState + 1)) {
+        return
+      }
+    }
+  }
+
+  /**
+   * Release an acquired shared lock.
+   */
+  def releaseShared(inst: MapHolder): Unit = {
+    var lockState = 0
+    do {
+      lockState = UnsafeUtils.getIntVolatile(inst, lockStateOffset)
+    } while (!UnsafeUtils.unsafe.compareAndSwapInt(inst, lockStateOffset, lockState, lockState - 1))
+  }
+
   // curIdx has to be initialized to one less than the starting logical index
-  // NOTE: This always fetches items using the official API.  THis is slower, but guarantees that no matter
+  // NOTE: This always fetches items using the official API.  This is slower, but guarantees that no matter
   // how slowly the iterator user pulls, it will always be pulling the right thing even if state/mapPtr changes.
   private class SortedIDMapElemIterator(inst: MapHolder, var curIdx: Int, continue: NativePointer => Boolean)
   extends ElementIterator {
+    var closed: Boolean = false
     var nextElem: NativePointer = _
+
+    val caller: Exception = new Exception()
```

Comment:
I don't see this used?  Not sure what the point of this is, and it uses up extra memory for each iteration.

Reply:
Oops. Mistake. It shouldn't be here.